beluga_amcl/src/amcl_node.cpp

```diff
@@ -852,13 +838,44 @@ void AmclNode::map_callback(nav_msgs::msg::OccupancyGrid::SharedPtr map)
     );
   } catch (const std::invalid_argument & error) {
     RCLCPP_ERROR(get_logger(), "Coudn't instantiate the particle filter: %s", error.what());
+  }
+  return nullptr;
+}
+
+void AmclNode::map_callback(nav_msgs::msg::OccupancyGrid::SharedPtr map)
+{
+  RCLCPP_INFO(get_logger(), "A new map was received");
+
+  if (particle_filter_ && get_parameter("first_map_only").as_bool()) {
+    RCLCPP_WARN(
+      get_logger(),
+      "Ignoring new map because the particle filter has already been initialized");
     return;
   }
 
-  if (last_known_estimate_.has_value() && !get_parameter("always_reset_initial_pose").as_bool()) {
-    const auto & [pose, covariance] = last_known_estimate_.value();
-    reinitialize_with_pose(pose, covariance);
-  } else if (get_parameter("set_initial_pose").as_bool()) {
+  const auto global_frame_id = get_parameter("global_frame_id").as_string();
+  if (map->header.frame_id != global_frame_id) {
+    RCLCPP_WARN(
+      get_logger(),
+      "Map frame \"%s\" doesn't match global frame \"%s\"",
+      map->header.frame_id.c_str(),
+      global_frame_id.c_str());
+  }
+
+  bool should_reset_initial_pose = true;
+  if (!particle_filter_) {
+    particle_filter_ = make_particle_filter(std::move(map));
+    if (last_known_estimate_.has_value() && !get_parameter("always_reset_initial_pose").as_bool()) {
+      const auto & [pose, covariance] = last_known_estimate_.value();
+      reinitialize_with_pose(pose, covariance);
+      should_reset_initial_pose = false;
+    }
+  } else {
+    particle_filter_->update_map(OccupancyGrid{std::move(map)});
+    should_reset_initial_pose = get_parameter("always_reset_initial_pose").as_bool();
+  }
+
+  if (should_reset_initial_pose && get_parameter("set_initial_pose").as_bool()) {
     const auto pose = Sophus::SE2d{
```

Comment:
when the node is deactivated the particleFilter save the state , which is different of the state which is save in the parameters. The question is when is better to use the particlefilter state, because in nav2_amcl the parameters are saved during deactivate state (in theory).

Reply:
Not sure what you mean @olmerg. We allow all combinations AFAICS.

Reply:
ohh I understand with `last_known_estimate_.has_value()` is covered the case, great :)